Backports for 1.11.3

KristofferC opened the pull request:
Backported PRs:
- [x] #56640 <!-- subtype: fast path for Type == TypeVar -->
- [x] #56789 <!-- Fix test report alignment -->
- [x] #56784 <!-- fix `exp(weirdNaN)` -->
- [x] #56317 <!-- Fix generate_precompile statement grouping & avoid defining new func -->
- [x] #56787 <!-- Fix partially_inline for unreachable -->
- [x] #56411 <!-- xref `UnionAll` in the doc string of `where` -->
- [x] #56772 <!-- docs: fix edge case in rational number conversion `float(a//b)` -->
- [x] #56842 <!-- dict docs: document that ordering of keys/values/pairs match iterate -->
- [x] #56793 <!-- Extend `Base.rationalize` instead of defining new function `Base.MathConstants.rationalize` -->
- [ ] #56264 <!-- inference: fix inference error from constructing invalid `TypeVar` -->
- [x] #56905 <!-- fix precompilation error printing if `CI` is set -->
- [x] #55871 <!-- lowering: don't reverse handler order in `(pop-handler-list ...)` -->
- [x] #56391 <!-- irinterp: set `IR_FLAG_REFINED` for narrowed `PhiNode`s -->
- [x] #56844 <!-- precompileplkgs: release parallel limiter when waiting for another process -->
- [x] #56644 <!-- fix jl_mutex_lock deadlock under rr -->
- [x] #56306 <!-- InteractiveUtils.jl: fixes issue where subtypes resolves bindings and causes deprecation warnings  -->
- [x] #56901 <!-- precompilepkgs: respect loaded dependencies when precompiling for load -->

Need manual backport:
- [ ] #56097 <!-- rearrange jl_delete_thread to be thread-safe -->
- [ ] #56329 <!-- loading: clean up more concurrency issues -->
- [ ] #56236 <!-- Allow taking Matrix slices without an extra allocation -->
- [ ] #55223 <!-- count bytes allocated through malloc more precisely -->
- [ ] #56749 <!-- precompile: don't waste memory on useless inferred code -->
- [ ] #56801 <!-- gc: improve mallocarrays locality -->

Contains multiple commits, manual intervention needed:
- [ ] #55348 <!-- Materialize complex Symmetric matrices in eigen -->
- [ ] #55877 <!-- fix FileWatching designs and add workaround for a stat bug on Apple -->
- [ ] #54457 <!-- Make `String(::Memory)` copy -->
- [ ] #42080 <!-- recommend explicit `using Foo: Foo, ...` in package code (was: "using considered harmful") -->
- [ ] #56625 <!-- codegen: make more judicious use of global rooting -->
- [ ] #56755 <!-- docs: fix scope type of a `struct` to hard -->
- [ ] #56856 <!-- fix abstract_eval_cfunction mistakes -->
- [x] #56914 <!-- Don't report only-inferred methods as recompiles -->

Non-merged PRs with backport label:
- [ ] #56915 <!-- Compiler: fix `tmerge(Const(s), Const(t))` st. `(s !== t) && (s == t)` -->
- [ ] #56895 <!-- REPL: Disable hints during non-bracketed paste -->
- [ ] #56890 <!-- Enable getting non-boxed LLVM type from Julia Type -->
- [ ] #56813 <!-- Miscellaneous `Meta.partially_inline!` fixes -->
- [ ] #56794 <!-- Make `Pairs` public -->
- [ ] #56709 <!-- Update version that version-named manifests works from -->
- [ ] #56281 <!-- REPL: complete imports when there are multiple macros before -->
- [ ] #55958 <!-- also redirect JL_STDERR etc. when redirecting to devnull -->
- [ ] #55956 <!-- Make threadcall gc safe -->
- [ ] #55813 <!-- Check for conflicting `@ccallable` name before JIT registration -->
- [ ] #55534 <!-- Set stdlib sources as read-only during installation -->
- [ ] #55499 <!-- propagate the terminal's `displaysize` to the `IOContext` used by the REPL -->
- [ ] #55458 <!-- Allow for generically extracting unannotated string -->
- [ ] #55457 <!-- Make AnnotateChar equality consider annotations -->
- [ ] #55220 <!-- `isfile_casesensitive` fixes on Windows -->
- [ ] #55169 <!-- `propertynames` for SVD respects private argument -->
- [ ] #53957 <!-- tweak how filtering is done for what packages should be precompiled -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->
- [ ] #50813 <!-- More doctests for Sockets and capitalization fix -->
- [ ] #50157 <!-- improve docs for `@inbounds` and `Base.@propagate_inbounds` --

gbaraldi committed:
Utilize bitshifts correctly in signals-mach.c when storing/reading the previous GC state (#53868)

I have not succeed in writing a test for this, but this was found on a
CI hang together with @keno and @vtjnash.

In essence if we hit a safepoint while GC_SAFE things can go wrong
<img width="748" alt="image"
src="https://github.com/JuliaLang/julia/assets/28694980/7d8170ee-11ab-43de-9bb1-9219aa5a2d80">

(cherry picked from commit 1e50a99fd6cdce69ed819b587709151df9d227a2)

vtjnash committed:
rearrange jl_delete_thread to be thread-safe (#56097)

Prior to this, especially on macOS, the gc-safepoint here would cause
the process to segfault as we had already freed the current_task state.
Rearrange this code so that the GC interactions (except for the atomic
store to current_task) are all handled before entering GC safe, and then
signaling the thread is deleted (via setting current_task = NULL,
published by jl_unlock_profile_wr to other threads) is last.

```
ERROR: Exception handler triggered on unmanaged thread.
Process 53827 stopped
* thread #5, stop reason = EXC_BAD_ACCESS (code=2, address=0x100018008)
    frame #0: 0x0000000100b74344 libjulia-internal.1.12.0.dylib`jl_delete_thread [inlined] jl_gc_state_set(ptls=0x000000011f8b3200, state='\x02', old_state=<unavailable>) at julia_threads.h:272:9 [opt]
   269 	    assert(old_state != JL_GC_CONCURRENT_COLLECTOR_THREAD);
   270 	    jl_atomic_store_release(&ptls->gc_state, state);
   271 	    if (state == JL_GC_STATE_UNSAFE || old_state == JL_GC_STATE_UNSAFE)
-> 272 	        jl_gc_safepoint_(ptls);
   273 	    return old_state;
   274 	}
   275 	STATIC_INLINE int8_t jl_gc_state_save_and_set(jl_ptls_t ptls,
Target 0: (julia) stopped.
(lldb) up
frame #1: 0x0000000100b74320 libjulia-internal.1.12.0.dylib`jl_delete_thread [inlined] jl_gc_state_save_and_set(ptls=0x000000011f8b3200, state='\x02') at julia_threads.h:278:12 [opt]
   275 	STATIC_INLINE int8_t jl_gc_state_save_and_set(jl_ptls_t ptls,
   276 	                                              int8_t state)
   277 	{
-> 278 	    return jl_gc_state_set(ptls, state, jl_atomic_load_relaxed(&ptls->gc_state));
   279 	}
   280 	#ifdef __clang_gcanalyzer__
   281 	// these might not be a safepoint (if they are no-op safe=>safe transitions), but we have to assume it could be (statically)
(lldb)
frame #2: 0x0000000100b7431c libjulia-internal.1.12.0.dylib`jl_delete_thread(value=0x000000011f8b3200) at threading.c:537:11 [opt]
   534 	    ptls->root_task = NULL;
   535 	    jl_free_thread_gc_state(ptls);
   536 	    // then park in safe-region
-> 537 	    (void)jl_gc_safe_enter(ptls);
   538 	}
```

(test incorporated into https://github.com/JuliaLang/julia/pull/55793)

(cherry picked from commit 0d09f3d0baa8843bd41fed013dad3fd9d69ae28d,
resolving conflicts from not having backported https://github.com/JuliaLang/julia/pull/52198)

aviatesk committed:
inference: fix inference error from constructing invalid `TypeVar` (#56264)

- fixes JuliaLang/julia#56248

(cherry picked from commit 08d11d041b22fe90380e56be4fb4d44aaf46ec85)

dgleich committed:
InteractiveUtils.jl: fixes issue where subtypes resolves bindings and causes deprecation warnings  (#56306)

The current version of `subtypes` will throw deprecation errors even if
no one is using the deprecated bindings.

A similar bug was fixed in Aqua.jl -
https://github.com/JuliaTesting/Aqua.jl/pull/89/files

See discussion here:

- https://github.com/JuliaIO/ImageMagick.jl/issues/235 (for identifying
the problem)
- https://github.com/simonster/Reexport.jl/issues/42 (for pointing to
the issue in Aqua.jl)
- https://github.com/JuliaTesting/Aqua.jl/pull/89/files (for the fix in
Aqua.jl)

This adds the `isbindingresolved` test to the `subtypes` function to
avoid throwing deprecation warnings. It also adds a test to check that
this doesn't happen.

---

On the current master branch (before the fix), the added test shows:

```
WARNING: using deprecated binding InternalModule.MyOldType in OuterModule.
, use MyType instead.
Subtypes and deprecations: Test Failed at /home/dgleich/devextern/julia/usr/share/julia/stdlib/v1.12/Test/src/Test.jl:932
  Expression: isempty(stderr_content)
   Evaluated: isempty("WARNING: using deprecated binding InternalModule.MyOldType in OuterModule.\n, use MyType instead.\n")
Test Summary:             | Fail  Total  Time
Subtypes and deprecations |    1      1  2.8s
ERROR: LoadError: Some tests did not pass: 0 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/dgleich/devextern/julia/stdlib/InteractiveUtils/test/runtests.jl:841
ERROR: Package InteractiveUtils errored during testing
```

---

Using the results of this pull request:

```
@test_nowarn subtypes(Integer);
```

passes without error. The other tests pass too.

(cherry picked from commit 20f933ae785423d3b94474e11453794c1fadc188)

aviatesk committed:
irinterp: set `IR_FLAG_REFINED` for narrowed `PhiNode`s (#56391)

`adce_pass!` can transform a `Union`-type `PhiNode` into a narrower
`PhiNode`, but in such cases, the `IR_FLAG_REFINED` flag isn’t set on
that `PhiNode` statement. By setting this flag, irinterp can perform
statement reprocessing using the narrowed `PhiNode`, enabling type
stability in cases like JuliaLang/julia#56387.

- fixes JuliaLang/julia#56387

(cherry picked from commit 671e1d87595d706ec2fa1659d990c507596f8fbc)

topolarity committed:
lowering: don't reverse handler order in `(pop-handler-list ...)` (#55871)

We were accidentally emitting a different pop order for `Expr(:leave,
...)` if you uncomment the `nothing` below:
```julia
let src = Meta.@lower let
    try
        try
            return 1
        catch
        end
    finally
        # nothing # <- uncomment me
    end
end
    println.(filter(stmt->Base.isexpr(stmt, :leave), src.args[1].code))
    nothing
end
```

(cherry picked from commit deac82ad915db6c1fd6e9246fd4bb3310b7cd71f)

vtjnash committed:
fix jl_mutex_lock deadlock under rr (#56644)

(cherry picked from commit bdf78c9ece6f46d71ef78801410deb6fe99af642)

N5N3 committed:
subtype: fast path for Type == TypeVar (#56640)

close #56606

(cherry picked from commit 7df1dfa2f1a51755eb35a2b5e51c810674dc10ee)

vtjnash committed:
precompile: don't waste memory on useless inferred code (#56749)

We never have a reason to reference this data again since we already
have native code generated for it, so it is simply wasting memory and
download space.

$ du -sh {old,new}/usr/share/julia/compiled
256M old
227M new

(cherry picked from commit dfe6a13e5038c8cbe0f1720d190629225ec1a19b)

IanButterworth committed:
Fix test report alignment (#56789)

(cherry picked from commit 32ea18ecfb545c6abae8013d133d683987b9e323)

maleadt commented:
This also needs to include LinearAlgebra.jl back-ports: https://github.com/JuliaLang/LinearAlgebra.jl/issues?q=state%3Aclosed%20label%3A%22backport%201.11%22
Or should those be made into commits here against `release-1.11` (where the stdlib hasn't been moved out)?

KristofferC commented:
> Or should those be made into commits here against release-1.11 (where the stdlib hasn't been moved out)?

I've been thinking about it and I think the easiest is just to cherry pick these commits onto the julia repo onto either a `backports-linearalgebra-1.11` branch or directly to `backports-1.11`.

oscardssmith committed:
fix `exp(weirdNaN)` (#56784)

Fixes https://github.com/JuliaLang/julia/issues/56782
Fix `exp(reinterpret(Float64, 0x7ffbb14880000000))` returning non-NaN
value. This should have minimal performance impact since it's already in
the fallback branch.

(cherry picked from commit 19e06d39141e1e2f0f9c4664ba9b88867b587a6e)

IanButterworth committed:
Fix generate_precompile statement grouping & avoid defining new func (#56317)

(cherry picked from commit 582585bfb0a5494d41fd28b47a9774d498c608d9)

IanButterworth commented:
@nanosoldier `runtests(vs=“release-1.10”)`

wsmoses committed:
Fix partially_inline for unreachable (#56787)

(cherry picked from commit 9118ea7565feae13d5b47654a3c245c0df36e753)

maleadt commented:
@IanButterworth You typo'd the syntax, branches need to be specified using `:`: https://github.com/JuliaCI/Nanosoldier.jl?tab=readme-ov-file#trigger-syntax. In any case, it shouldn't be required here, as PkgEval uses the merge-base, which should be `release-1.10` already.

@nanosoldier `runtests()`

IanButterworth commented:
Ah ok, but this is a 1.11 backports PR. I wanted to check it against 1.10

DilumAluthge commented:
I don't see the Nanosoldier commit status here, which leads me to believe that either the webhook was never delivered, or that the Nanosoldier server didn't accept the job.

DilumAluthge commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56741#issuecomment-2540944084) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c3449dc_vs_5e9a32e/report.html) is available.

IanButterworth commented:
@nanosoldier `runtests(vs=":release-1.10")`

IanButterworth commented:
@maleadt please cancel the duplicate run invoked here https://github.com/JuliaLang/julia/pull/56741#issuecomment-2542521100

The latest request I did is the correct spec. We want to check against 1.10 state

maleadt commented:
> please cancel the duplicate run invoked here [#56741 (comment)](https://github.com/JuliaLang/julia/pull/56741#issuecomment-2542521100) (or let me know how to if I can)

It is currently not possible to cancel Nanosoldier invocations. Only thing I can to is restart the service at a strategic time. I'll check tomorrow when this job starts.

nsajko committed:
xref `UnionAll` in the doc string of `where` (#56411)

(cherry picked from commit 8924e6f6c4f61946e72fde783c59a7996d749051)

maleadt commented:
@nanosoldier `runtests(vs=":release-1.10")`

maleadt commented:
@nanosoldier `runtests(vs=":release-1.10")`

Commit:
bump Pkg to latest 1.11

maleadt commented:
@nanosoldier `runtests(vs=":release-1.10")`

jakobnissen commented:
The ScanByte segfault is a package issue and not a problem with Julia 1.11. I've added it to the PkgEval blacklist.

IanButterworth committed:
precompileplkgs: release parallel limiter when waiting for another process (#56844)

(cherry picked from commit 37740148e1120ec833d6648dde245c0c5b79d76e)

IanButterworth commented:
I haven't looked into the [pkgeval run against 1.10 in detail ](https://github.com/JuliaLang/julia/pull/56741#issuecomment-2547941064) but:

> 593 packages failed only on the current version 
> - Package fails to precompile (99 packages):
> - Illegal method overwrites during precompilation (10 packages):
> - Package has test failures (128 packages):
> - Package tests unexpectedly errored (126 packages):
> - Package is missing a package dependency (1 packages):
> - Package is using an unknown package (2 packages):
> - There were unidentified errors (19 packages):
> - Tests became inactive (3 packages):
> - Test duration exceeded the time limit (203 packages):
> - Test log exceeded the size limit (2 packages):


> 191 packages passed tests only on the current version.

Commit:
docs: fix edge case in rational number conversion `float(a//b)` (#56772)

Fixes #56726 added the changes that were suggested. fixing the mistake.

---------

Co-authored-by: Max Horn <max@quendi.de>
Co-authored-by: Chengyu Han <git@wo-class.cn>
(cherry picked from commit fc9e7c4a6a36d7d54ae7dd8b89ac14a8334a59c4)

cossio committed:
dict docs: document that ordering of keys/values/pairs match iterate (#56842)

Fix #56841.

Currently the documentation states that keys(dict) and values(dict)
iterate in the same order. But it is not stated whether this is the same
order as that used by pairs(dict), or when looping, for (k,v) in dict.

This PR makes this guarantee explicit.

(cherry picked from commit 796d823b4fbb3ec94c09f7022f9b614fc846dced)

sostock committed:
Extend `Base.rationalize` instead of defining new function (#56793)

#55886 accidentally created a new function
`Base.MathConstants.rationalize` instead of extending
`Base.rationalize`, which is the reason why `Base.rationalize(Int, π)`
isn’t constant-folded in Julia 1.10 and 1.11:

```
julia> @btime rationalize(Int,π);
  1.837 ns (0 allocations: 0 bytes)      # v1.9: constant-folded
  88.416 μs (412 allocations: 15.00 KiB) # v1.10: not constant-folded
```

This PR fixes that. It should probably be backported to 1.10 and 1.11.

(cherry picked from commit d3c26b7ecce7e8319788bc0a035241d6dfd21ac0)

simeonschaub committed:
fix precompilation error printing if `CI` is set (#56905)

Co-authored-by: Ian Butterworth <i.r.butterworth@gmail.com>
(cherry picked from commit e3f90c4062af3a40ee548203242744424f2029fa)

IanButterworth committed:
precompilepkgs: respect loaded dependencies when precompiling for load (#56901)

(cherry picked from commit 27ebbf01ccc44d78694516ff6daf55af65b7439b)

Commit:
avoid memory leak in mallocarrays linked list

IanButterworth committed:
Don't report only-inferred methods as recompiles (#56914)

IanButterworth committed:
REPL: Limit method lookup when completing kwargs (#56963)

(cherry picked from commit 10e20f12679a549d82337af14a2caa709e03b88b)

KristofferC committed:
fix backport of #56097

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

giordano committed:
[BinaryPlatforms] Parse `rc64`/`riscv64` triplets

This is partial backport of #56105, only the part relative to parsing triplets.

giordano commented:
I pushed https://github.com/JuliaLang/julia/commit/cbb7ebc257e28ddb15af21f34b6facc35c199909 to make it possible to parse `riscv64` triplets.  I verified this solves issues like https://discourse.julialang.org/t/dsp-jl-loading-error/124345 for me locally, by installing `JLLWrappers@1.6` in the global environment, using it, then activating another environment, installing `HelloWorldC_jll@1.4` in there and using that package: on Julia v1.11.2 that fails wit the error reported on discourse, on cbb7ebc257e28ddb15af21f34b6facc35c199909 no error is raised (although it remains the fact the wrong version of `JLLWrappers` is actually loaded).

willtebbutt committed:
fix deepcopy for non-trivial circular references (#56990)

Resolves #56775 . Credit for this fix rests entirely with bbrehm .

---------

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>
Co-authored-by: Neven Sajko <4944410+nsajko@users.noreply.github.com>
(cherry picked from commit 1ebacacc5a82e16930bd5ae70b641a3cbd6ee883)

nsajko committed:
broadcast: align `ndims` implementation with intent behind code (#56999)

The `N<:Integer` constraint was nonsensical, given that `(N === Any) ||
(N isa Int)`. N5N3 noticed this back in 2022:
https://github.com/JuliaLang/julia/pull/44061#discussion_r883261337

Follow up on #44061. Also xref #45477.

(cherry picked from commit d3964b600a5e596bd47cf04b3da4a1e78d4aecaf)

DilumAluthgeBot committed:
🤖 [backports-release-1.11] Bump the Pkg stdlib from bac7775db to 2eb8ae5b8 (#57015)

Co-authored-by: IanButterworth <1694067+IanButterworth@users.noreply.github.com>

topolarity committed:
Miscellaneous `Meta.partially_inline!` fixes (#56813)

This fixes up a couple of conspicuous problems I noticed when reviewing
#56787

I'm unsure we should have accepted this pass in Base to begin with... It
should at least be re-written to error on unexpected IR elements
(instead of performing an invalid transform silently), but the real
problem is that this pass is out-of-pipeline and so it doesn't run on
most user code and we have much worse coverage compared to passes in the
main Compiler.

(cherry picked from commit c1db3a419699a92c29ce2aff7547019eec3045a4)

KristofferC commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56741#issuecomment-2587038498) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/0aa7b95_vs_5e9a32e/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - GC corruption was detected: 1 packages

28 packages crashed on the previous version too.

## ✖ Packages that failed

**34 packages failed only on the current version.**

 - Illegal method overwrites during precompilation: 1 packages
 - Package tests unexpectedly errored: 7 packages
 - Test duration exceeded the time limit: 25 packages
 - Test log exceeded the size limit: 1 packages

2792 packages failed on the previous version too.

## ✔ Packages that passed tests

**41 packages passed tests only on the current version.**

 - Other: 41 packages

6044 packages passed tests on the previous version too.

## ➖ Packages that were skipped altogether

1309 packages were skipped on the previous version too.



</details>

KristofferC committed:
fix kwarg limit method backport

KristofferC commented:
@nanosoldier `runtests(["NumericalAlgorithms", "JuliaInterpreter", "LoopManagers", "DiffusionMap", "CharacteristicInvFourier", "Intervals", "GLPK", "MRICoilSensitivities", "GaussianMixtureRegressions", "GaussianMixtures", "ClimaAnalysis", "TableTransforms", "OptimalControl", "Braket", "FSimBase", "OptimizationManopt", "Surrogates", "NonconvexMultistart", "CompressiveLearning", "KernelInterpolation", "UltraDark", "StructuralEquationModels", "Knockoffs", "GeoStatsValidation", "LiquidElectrolytes", "PhyloClustering", "BoxCox", "QuadraticProgramNetworks", "CDGRNs", "BloqadeODE", "REPL", "PyramidScheme"])`

KristofferC committed:
Revert "Avoid allocations in views of views (#53231)" (#57044)

This reverts commit 2bd4cf8090f2c651543f562a4c469d73e4b15bd6. (#53231)

The reason for this revert is that it caused exponential blowup of types
in iterated views causing some packages to simply freeze when doing
something that worked ok in 1.10.

In my opinion, the perf gain from the PR is not outweighed by the "risk"
of hitting this compilation blowup case.

Fixes https://github.com/JuliaLang/julia/issues/56760.

Co-authored-by: KristofferC <kristoffer.carlsson@juliacomputing.com>
(cherry picked from commit b23f557e80cabac1624fc98abacfba152156ef8f)

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/56741#issuecomment-2589883558) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f66a1cf_vs_5e9a32e/report.html) is available.

<details><summary>Report summary</summary>

## ✖ Packages that failed

**7 packages failed only on the current version.**

 - Package tests unexpectedly errored: 1 packages
 - Test duration exceeded the time limit: 6 packages

2 packages failed on the previous version too.

## ✔ Packages that passed tests

23 packages passed tests on the previous version too.



</details>
